kv/kvserver: TestStoreCapacityAfterSplit failed

kv/kvserver.TestStoreCapacityAfterSplit [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/17365965?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/17365965?buildTab=artifacts#/) on release-23.1 @ [b24fcc7fb7f334f3b7fad9cc7bfbe67e4b9e515f](https://github.com/cockroachdb/cockroach/commits/b24fcc7fb7f334f3b7fad9cc7bfbe67e4b9e515f):


```
=== RUN   TestStoreCapacityAfterSplit
    test_log_scope.go:156: test logs captured to: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestStoreCapacityAfterSplit3875734265
    test_log_scope.go:74: use -show-logs to present logs inline
    client_split_test.go:2866: [NotLeaseHolderError] stale proposal: command was proposed under lease #2 but is being applied under lease: repl=(n2,s2):2 seq=3 start=1729255736.499453783,0 epo=1 pro=1729255736.500969625,0; r64: replica not lease holder; current lease is repl=(n2,s2):2 seq=3 start=1729255736.499453783,0 epo=1 pro=1729255736.500969625,0
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33e1d369c27b9c01b2b6009c561815a3/logTestStoreCapacityAfterSplit3875734265
--- FAIL: TestStoreCapacityAfterSplit (1.06s)
```

Parameters:
 - <code>TAGS=bazel,gss</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStoreCapacityAfterSplit.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-43344

--- a/pkg/kv/kvserver/client_split_test.go
+++ b/pkg/kv/kvserver/client_split_test.go
@@ -3267,7 +3267,23 @@ func TestStoreCapacityAfterSplit(t *testing.T) {
 	key := tc.ScratchRange(t)
 	desc := tc.AddVotersOrFatal(t, key, tc.Target(1))
 	tc.TransferRangeLeaseOrFatal(t, desc, tc.Target(1))
-	tc.WaitForLeaseUpgrade(ctx, t, desc)
+
+	// Wait for the lease transfer to be applied on the new leaseholder and then
+	// to be upgraded from an expiration-based lease.
+	testutils.SucceedsSoon(t, func() error {
+		repl, err := s.GetReplica(desc.RangeID)
+		if err != nil {
+			return err
+		}
+		l, _ := repl.GetLease()
+		if !l.OwnedBy(s.StoreID()) {
+			return errors.Errorf("lease transfer not applied on leaseholder")
+		}
+		if l.Type() == roachpb.LeaseExpiration {
+			return errors.Errorf("lease still an expiration based lease")
+		}
+		return nil
+	})
 
 	cap, err := s.Capacity(ctx, false /* useCached */)
 	if err != nil {

